File: sdk-typescript/packages/editor/src/editor.ts
/**
 * WAVE SDK - Editor API
 *
 * Video editing capabilities including cuts, transitions, overlays, and effects.
 *
 * NOTE: This is a client SDK. All authorization checks are performed server-side.
 * The API will return 403 Forbidden if the user lacks required permissions.
 */

import type {
  WaveClient,
  PaginationParams,
  PaginatedResponse,
} from '@wave-av/core';

// Types extracted to editor-types.ts — re-export all for backwards compat
export * from './editor-types';
import type {
  AddElementRequest,
  CreateProjectRequest,
  EditorProject,
  Effect,
  ListProjectsParams,
  RenderJob,
  RenderOptions,
  TimelineElement,
  Track,
  TrackType,
  Transition,
  UpdateProjectRequest,
} from './editor-types';
export type { ProjectStatus, TrackType, TransitionType, EffectType, TimelineElement, Track, Transition, Effect, Keyframe, TextOverlay, TextAnimation, EditorProject, CreateProjectRequest, UpdateProjectRequest, AddElementRequest, RenderOptions, RenderJob, ListProjectsParams } from './editor-types';

export class EditorAPI {
  private readonly client: WaveClient;
  private readonly basePath = '/v1/editor/projects';

  constructor(client: WaveClient) {
    this.client = client;
  }

  /**
   * Create a new editor project
   *
   * Requires: editor:create permission
   */
  async createProject(request: CreateProjectRequest): Promise<EditorProject> {
    return this.client.post<EditorProject>(this.basePath, request);
  }

  /**
   * Get a project by ID
   *
   * Requires: editor:read permission
   */
  async getProject(projectId: string): Promise<EditorProject> {
    return this.client.get<EditorProject>(`${this.basePath}/${projectId}`);
  }

  /**
   * Update a project
   *
   * Requires: editor:update permission
   */
  async updateProject(
    projectId: string,
    request: UpdateProjectRequest
  ): Promise<EditorProject> {
    return this.client.patch<EditorProject>(`${this.basePath}/${projectId}`, request);
  }

  /**
   * Remove a project
   *
   * Requires: editor:remove permission (server-side RBAC enforced)
   */
  async removeProject(projectId: string): Promise<void> {
    await this.client.delete(`${this.basePath}/${projectId}`);
  }

  /**
   * List projects
   *
   * Requires: editor:read permission
   */
  async listProjects(params?: ListProjectsParams): Promise<PaginatedResponse<EditorProject>> {
    return this.client.get<PaginatedResponse<EditorProject>>(this.basePath, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }

  /**
   * Duplicate a project
   *
   * Requires: editor:create permission
   */
  async duplicateProject(projectId: string, name?: string): Promise<EditorProject> {
    return this.client.post<EditorProject>(`${this.basePath}/${projectId}/duplicate`, {
      name,
    });
  }

  /**
   * Add a track to a project
   *
   * Requires: editor:update permission
   */
  async addTrack(
    projectId: string,
    track: { name: string; type: TrackType; order?: number }
  ): Promise<Track> {
    return this.client.post<Track>(`${this.basePath}/${projectId}/tracks`, track);
  }

  /**
   * Update a track
   *
   * Requires: editor:update permission
   */
  async updateTrack(
    projectId: string,
    trackId: string,
    updates: Partial<Pick<Track, 'name' | 'order' | 'locked' | 'muted' | 'visible'>>
  ): Promise<Track> {
    return this.client.patch<Track>(
      `${this.basePath}/${projectId}/tracks/${trackId}`,
      updates
    );
  }

  /**
   * Remove a track
   *
   * Requires: editor:update permission (server-side RBAC enforced)
   */
  async removeTrack(projectId: string, trackId: string): Promise<void> {
    await this.client.delete(
      `${this.basePath}/${projectId}/tracks/${trackId}`,
      { method: 'DELETE' }
    );
  }

  /**
   * Add an element to a track
   *
   * Requires: editor:update permission
   */
  async addElement(projectId: string, element: AddElementRequest): Promise<TimelineElement> {
    return this.client.post<TimelineElement>(
      `${this.basePath}/${projectId}/elements`,
      element
    );
  }

  /**
   * Update an element
   *
   * Requires: editor:update permission
   */
  async updateElement(
    projectId: string,
    elementId: string,
    updates: Partial<TimelineElement>
  ): Promise<TimelineElement> {
    return this.client.patch<TimelineElement>(
      `${this.basePath}/${projectId}/elements/${elementId}`,
      updates
    );
  }

  /**
   * Remove an element
   *
   * Requires: editor:update permission (server-side RBAC enforced)
   */
  async removeElement(projectId: string, elementId: string): Promise<void> {
    await this.client.delete(
      `${this.basePath}/${projectId}/elements/${elementId}`,
      { method: 'DELETE' }
    );
  }

  /**
   * Move an element to a different position
   *
   * Requires: editor:update permission
   */
  async moveElement(
    projectId: string,
    elementId: string,
    options: { track_id?: string; start_time: number }
  ): Promise<TimelineElement> {
    return this.client.post<TimelineElement>(
      `${this.basePath}/${projectId}/elements/${elementId}/move`,
      options
    );
  }

  /**
   * Trim an element
   *
   * Requires: editor:update permission
   */
  async trimElement(
    projectId: string,
    elementId: string,
    options: { in_point?: number; out_point?: number }
  ): Promise<TimelineElement> {
    return this.client.post<TimelineElement>(
      `${this.basePath}/${projectId}/elements/${elementId}/trim`,
      options
    );
  }

  /**
   * Add a transition between elements
   *
   * Requires: editor:update permission
   */
  async addTransition(
    projectId: string,
    transition: Omit<Transition, 'id'>
  ): Promise<Transition> {
    return this.client.post<Transition>(
      `${this.basePath}/${projectId}/transitions`,
      transition
    );
  }

  /**
   * Update a transition
   *
   * Requires: editor:update permission
   */
  async updateTransition(
    projectId: string,
    transitionId: string,
    updates: Partial<Transition>
  ): Promise<Transition> {
    return this.client.patch<Transition>(
      `${this.basePath}/${projectId}/transitions/${transitionId}`,
      updates
    );
  }

  /**
   * Remove a transition
   *
   * Requires: editor:update permission (server-side RBAC enforced)
   */
  async removeTransition(projectId: string, transitionId: string): Promise<void> {
    await this.client.delete(
      `${this.basePath}/${projectId}/transitions/${transitionId}`,
      { method: 'DELETE' }
    );
  }

  /**
   * Add an effect to an element
   *
   * Requires: editor:update permission
   */
  async addEffect(projectId: string, effect: Omit<Effect, 'id'>): Promise<Effect> {
    return this.client.post<Effect>(
      `${this.basePath}/${projectId}/effects`,
      effect
    );
  }

  /**
   * Update an effect
   *
   * Requires: editor:update permission
   */
  async updateEffect(
    projectId: string,
    effectId: string,
    updates: Partial<Effect>
  ): Promise<Effect> {
    return this.client.patch<Effect>(
      `${this.basePath}/${projectId}/effects/${effectId}`,
      updates
    );
  }

  /**
   * Remove an effect
   *
   * Requires: editor:update permission (server-side RBAC enforced)
   */
  async removeEffect(projectId: string, effectId: string): Promise<void> {
    await this.client.delete(
      `${this.basePath}/${projectId}/effects/${effectId}`,
      { method: 'DELETE' }
    );
  }

  /**
   * Start rendering a project
   *
   * Requires: editor:render permission
   */
  async render(projectId: string, options?: RenderOptions): Promise<RenderJob> {
    return this.client.post<RenderJob>(
      `${this.basePath}/${projectId}/render`,
      options
    );
  }

  /**
   * Get render job status
   *
   * Requires: editor:read permission
   */
  async getRenderJob(projectId: string, jobId: string): Promise<RenderJob> {
    return this.client.get<RenderJob>(
      `${this.basePath}/${projectId}/render/${jobId}`
    );
  }

  /**
   * List render jobs for a project
   *
   * Requires: editor:read permission
   */
  async listRenderJobs(
    projectId: string,
    params?: PaginationParams
  ): Promise<PaginatedResponse<RenderJob>> {
    return this.client.get<PaginatedResponse<RenderJob>>(
      `${this.basePath}/${projectId}/render`,
      { params: params as Record<string, string | number | boolean | undefined> }
    );
  }

  /**
   * Cancel a render job
   *
   * Requires: editor:render permission
   */
  async cancelRenderJob(projectId: string, jobId: string): Promise<RenderJob> {
    return this.client.post<RenderJob>(
      `${this.basePath}/${projectId}/render/${jobId}/cancel`
    );
  }

  /**
   * Wait for render to complete
   */
  async waitForRender(
    projectId: string,
    jobId: string,
    options?: {
      pollInterval?: number;
      timeout?: number;
      onProgress?: (job: RenderJob) => void;
    }
  ): Promise<RenderJob> {
    const pollInterval = options?.pollInterval || 3000;
    const timeout = options?.timeout || 1800000; // 30 minutes
    const startTime = Date.now();

    while (Date.now() - startTime < timeout) {
      const job = await this.getRenderJob(projectId, jobId);

      if (options?.onProgress) {
        options.onProgress(job);
      }

      if (job.status === 'ready') {
        return job;
      }

      if (job.status === 'failed') {
        throw new Error(`Render failed: ${job.error || 'Unknown error'}`);
      }

      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }

    throw new Error(`Render timed out after ${timeout}ms`);
  }

  /**
   * Generate a preview frame
   *
   * Requires: editor:read permission
   */
  async getPreviewFrame(
    projectId: string,
    time: number,
    options?: { width?: number; height?: number; format?: 'png' | 'jpeg' }
  ): Promise<{ url: string; expires_at: string }> {
    return this.client.get(`${this.basePath}/${projectId}/preview`, {
      params: { time, ...options } as unknown as Record<string, string | number | boolean | undefined>,
    });
  }

  /**
   * Generate a preview video segment
   *
   * Requires: editor:read permission
   */
  async getPreviewSegment(
    projectId: string,
    startTime: number,
    endTime: number,
    options?: { quality?: 'low' | 'medium' }
  ): Promise<{ url: string; expires_at: string }> {
    return this.client.post(`${this.basePath}/${projectId}/preview/segment`, {
      start_time: startTime,
      end_time: endTime,
      ...options,
    });
  }
}

/**
 * Create an Editor API instance
 */
export function createEditorAPI(client: WaveClient): EditorAPI {
  return new EditorAPI(client);
}

File: sdk-python/wave/editor.py
"""WAVE SDK - Editor API. Video editing with tracks, transitions, effects, and rendering."""
from __future__ import annotations
import time
from typing import Any
from pydantic import BaseModel
from wave.client import WaveClient

class EditorProject(BaseModel):
    id: str; organization_id: str; title: str; status: str; duration: float = 0; track_count: int = 0; created_at: str; updated_at: str

class Track(BaseModel):
    id: str; project_id: str; name: str; type: str; sort_order: int = 0; muted: bool = False; locked: bool = False

class TimelineElement(BaseModel):
    id: str; track_id: str; type: str; source_url: str | None = None; start_time: float; end_time: float; in_point: float = 0; out_point: float = 0

class RenderJob(BaseModel):
    id: str; project_id: str; status: str; progress_percent: int = 0; output_url: str | None = None; error: str | None = None; created_at: str

class EditorAPI:
    def __init__(self, client: WaveClient): self._client = client; self._base = "/v1/editor/projects"
    def create_project(self, title: str, **kwargs: Any) -> EditorProject: return EditorProject(**self._client.post(self._base, json={"title": title, **kwargs}))
    def get_project(self, project_id: str) -> EditorProject: return EditorProject(**self._client.get(f"{self._base}/{project_id}"))
    def update_project(self, project_id: str, **kwargs: Any) -> EditorProject: return EditorProject(**self._client.patch(f"{self._base}/{project_id}", json=kwargs))
    def remove_project(self, project_id: str) -> None: self._client.delete(f"{self._base}/{project_id}")
    def list_projects(self, **params: Any) -> dict: return self._client.get(self._base, params={k: v for k, v in params.items() if v is not None})
    def duplicate_project(self, project_id: str) -> EditorProject: return EditorProject(**self._client.post(f"{self._base}/{project_id}/duplicate"))
    def add_track(self, project_id: str, name: str, type: str = "video", **kwargs: Any) -> Track: return Track(**self._client.post(f"{self._base}/{project_id}/tracks", json={"name": name, "type": type, **kwargs}))
    def update_track(self, project_id: str, track_id: str, **kwargs: Any) -> Track: return Track(**self._client.patch(f"{self._base}/{project_id}/tracks/{track_id}", json=kwargs))
    def remove_track(self, project_id: str, track_id: str) -> None: self._client.delete(f"{self._base}/{project_id}/tracks/{track_id}")
    def add_element(self, project_id: str, track_id: str, type: str, start_time: float, end_time: float, **kwargs: Any) -> TimelineElement: return TimelineElement(**self._client.post(f"{self._base}/{project_id}/tracks/{track_id}/elements", json={"type": type, "start_time": start_time, "end_time": end_time, **kwargs}))
    def update_element(self, project_id: str, element_id: str, **kwargs: Any) -> TimelineElement: return TimelineElement(**self._client.patch(f"{self._base}/{project_id}/elements/{element_id}", json=kwargs))
    def remove_element(self, project_id: str, element_id: str) -> None: self._client.delete(f"{self._base}/{project_id}/elements/{element_id}")
    def move_element(self, project_id: str, element_id: str, track_id: str, start_time: float) -> TimelineElement: return TimelineElement(**self._client.post(f"{self._base}/{project_id}/elements/{element_id}/move", json={"track_id": track_id, "start_time": start_time}))
    def trim_element(self, project_id: str, element_id: str, in_point: float, out_point: float) -> TimelineElement: return TimelineElement(**self._client.post(f"{self._base}/{project_id}/elements/{element_id}/trim", json={"in_point": in_point, "out_point": out_point}))
    def add_transition(self, project_id: str, type: str, element_a_id: str, element_b_id: str, duration_ms: int = 500) -> dict: return self._client.post(f"{self._base}/{project_id}/transitions", json={"type": type, "element_a_id": element_a_id, "element_b_id": element_b_id, "duration_ms": duration_ms})
    def update_transition(self, project_id: str, transition_id: str, **kwargs: Any) -> dict: return self._client.patch(f"{self._base}/{project_id}/transitions/{transition_id}", json=kwargs)
    def remove_transition(self, project_id: str, transition_id: str) -> None: self._client.delete(f"{self._base}/{project_id}/transitions/{transition_id}")
    def add_effect(self, project_id: str, element_id: str, type: str, **kwargs: Any) -> dict: return self._client.post(f"{self._base}/{project_id}/elements/{element_id}/effects", json={"type": type, **kwargs})
    def update_effect(self, project_id: str, effect_id: str, **kwargs: Any) -> dict: return self._client.patch(f"{self._base}/{project_id}/effects/{effect_id}", json=kwargs)
    def remove_effect(self, project_id: str, effect_id: str) -> None: self._client.delete(f"{self._base}/{project_id}/effects/{effect_id}")
    def get_preview_frame(self, project_id: str, time_seconds: float) -> dict: return self._client.get(f"{self._base}/{project_id}/preview/frame", params={"time": time_seconds})
    def get_preview_segment(self, project_id: str, start: float, end: float) -> dict: return self._client.get(f"{self._base}/{project_id}/preview/segment", params={"start": start, "end": end})
    def render(self, project_id: str, **kwargs: Any) -> RenderJob: return RenderJob(**self._client.post(f"{self._base}/{project_id}/render", json=kwargs if kwargs else None))
    def get_render_job(self, job_id: str) -> RenderJob: return RenderJob(**self._client.get(f"/v1/editor/renders/{job_id}"))
    def list_render_jobs(self, project_id: str, **params: Any) -> dict: return self._client.get(f"{self._base}/{project_id}/renders", params={k: v for k, v in params.items() if v is not None})
    def cancel_render_job(self, job_id: str) -> None: self._client.post(f"/v1/editor/renders/{job_id}/cancel")
    def wait_for_render(self, job_id: str, poll_interval: float = 3.0, timeout: float = 600.0) -> RenderJob:
        start = time.time()
        while time.time() - start < timeout:
            j = self.get_render_job(job_id)
            if j.status == "ready": return j
            if j.status == "failed": raise RuntimeError(f"Render failed: {j.error}")
            time.sleep(poll_interval)
        raise TimeoutError(f"Render timed out after {timeout}s")
